doc/sphinx/language/extensions/implicit-arguments.rst

```diff
@@ -203,6 +207,8 @@ For non maximally inserted implicit arguments, use square brackets:
 
    Print Implicit map.
 
+.. todo Can we give or better describe the full list of constructs?
```

Comment:
Well, that's pretty simple. The syntax is supported whenever a list of `binder` is expected.

Reply:
Is "all top-level definitions: :cmd:`Definition`, :cmd:`Fixpoint`, :cmd:`Lemma` and so on" the same as "in all function definitions" or "all top-level function definitions"?  If so, better to reference the defined term.

Of course, we need to add the definition somewhere and preferably a glossary index.

Reply:
It's not supported just in top-level function definitions. There's another paragraph below saying that it is also supported in internal binders. So basically, implicit binders are supported everywhere a binder is expected (i.e. they are supported wherever they are syntactically valid). I don't see why the text would have to say anything more complicated than that.

The example with the `forall` goes beyond "in all function definitions". In this case, the implicit binder is in the function type, `f` is not "defined", it is "assumed". A simpler example based on the same idea is:

```
Parameter f : forall {A}, A -> A * A.

Check (f 0).
```

Once again, `f` is "assumed" rather than "defined". It is its type rather than its definition that contains the implicit binder, and yet it is understood in the same way (you don't have to provide the implicit argument when using the function).

An interesting case is when you provide binders with conflicting status in a definition and in its type:

```coq
Definition f : forall {A}, A -> A * A := fun A x => (x, x).
(* Warning: Implicit arguments declaration relies on type.
Discarding incompatible declaration in term. [implicits-in-term,implicits] *)
Check (f 0).
Definition f : forall A, A -> A * A := fun {A} x => (x, x).
(* Warning: Implicit arguments declaration relies on type.
Discarding incompatible declaration in term. [implicits-in-term,implicits] *)
Check (f 0).
(* Error: The term "0" has type "nat" while it is expected to have type "Type". *)
```

Reply:
OK, that makes more sense.  The original wording is misleading if you try to dig into it

`The syntax is supported in all top-level definitions:
:cmd:`Definition`, :cmd:`Fixpoint`, :cmd:`Lemma` and so on.`

The section titles in the Gallina chapter use the terms "Assumptions", "Definitions" and "Assertions".  Do these have Coq-specific meaning (and should be defined) or do they just have their ordinary meaning?  We could be clearer about that.

The only commands under "Definitions" are `Definitions` and `Example`.  `Fixpoint` is under "Definition of recursive functions".  `Lemma` is under "Assertions".

I'll remove the sentence.

Reply:
I hope to clarify this soon while rewriting the Core language chapter, but indeed, we should generally consider two types of "definitions": transparent and opaque definitions (or to use the terminology of [CEP#42](https://github.com/coq/ceps/pull/42) to avoid the confusion with the `Transparent` and `Opaque` commands, "sealed" and "unsealed" definitions). "transparent" / "unsealed" definitions are most often simply called "definitions" and defined with `Definition` while "opaque" / "sealed" definitions are most often simply called "theorems" and defined with `Theorem` / `Lemma`. However, Coq allows to use the commands interchangeably.